src/Database/LSMTree/Internal/CRC32C.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Functionalty related to CRC-32C (Castagnoli) checksums:
+--
+-- * Support for calculating checksums while incrementally writing files.
+-- * Support for verifying checksums of files.
+-- * Support for a text file format listing file checksums.
+--
+module Database.LSMTree.Internal.CRC32C (
+
+  CRC32C(..),
+
+  -- * Pure incremental checksum calculation
+  initialCRC32C,
+  updateCRC32C,
+
+  -- * I\/O with checksum calculation
+  hGetSomeCRC32C,
+  hPutSomeCRC32C,
+  hPutAllCRC32C,
```

Comment:
Out of interest, what is the reason for not providing `hGetAllCRC32C`?

Reply:
I wasn't convinced that we'd need it, but we can add it if it turns out that we do. There isn't perfect symmetry. When reading a whole file sequentially and incrementally we have to deal with short reads anyway, due to the end of the file, so there doesn't seem to be any great benefit to something that tries a bit harder to read as much as you asked for.

A case where you would want to read exactly what was asked for is when we read a page, but that uses a completely different API and doesn't need to compute CRCs. We'll only be computing CRCs on writing whole files, and verifying them later. So that's sequential in both cases.